apps/openmw/mwsound/soundbufferpool.hpp

```diff
@@ -0,0 +1,145 @@
+            std::size_t mBufferCacheMin;
+            std::size_t mBufferCacheMax;
+            std::size_t mBufferCacheSize = 0;
+            std::list<Sound_Buffer*> mUnusedBuffers;
+            std::list<Sound_Buffer*> mUsedBuffers;
+
+            inline SoundBufferRef insertSound(const std::string& soundId, const ESM::Sound& sound);
+
+            inline void unloadUnused();
+    };
+
+    class SoundBufferSharedRef
+    {
+        public:
+            SoundBufferSharedRef(SoundBufferPool& pool, Sound_Buffer& value) noexcept
+                : mPool(&pool), mValue(&value)
+            {
+                mPool->acquire(*mValue);
+            }
+
+            SoundBufferSharedRef(const SoundBufferSharedRef& other) noexcept
+                : mPool(other.mPool), mValue(other.mValue)
+            {
+                mValue->incrementUses();
+            }
+
+            SoundBufferSharedRef(SoundBufferSharedRef&& other) noexcept
+                : mPool(other.mPool), mValue(other.mValue)
+            {
+                other.mPool = nullptr;
+                other.mValue = nullptr;
+            }
+
+            ~SoundBufferSharedRef() noexcept
+            {
+                if (mValue != nullptr)
+                    mPool->release(*mValue);
+            }
+
+            Sound_Buffer* get() const noexcept { return mValue; }
+
+            void reset() noexcept
+            {
+                if (mValue != nullptr)
+                {
+                    mPool->release(*mValue);
+                    mPool = nullptr;
+                    mValue = nullptr;
+                }
+            }
+
+            SoundBufferSharedRef& operator=(const SoundBufferSharedRef& other) noexcept
+            {
+                SoundBufferSharedRef copy(other);
+                std::swap(mPool, copy.mPool);
+                std::swap(mValue, copy.mValue);
+                return *this;
+            }
+
+            SoundBufferSharedRef& operator=(SoundBufferSharedRef&& other) noexcept
+            {
+                SoundBufferSharedRef moved(std::move(other));
+                std::swap(mPool, moved.mPool);
+                std::swap(mValue, moved.mValue);
+                return *this;
+            }
+
+        private:
+            SoundBufferPool* mPool = nullptr;
+            Sound_Buffer* mValue = nullptr;
+    };
+
+    class SoundBufferRef
+    {
+        public:
+            SoundBufferRef() = default;
+
+            SoundBufferRef(SoundBufferPool& pool, Sound_Buffer& value) noexcept
+                : mPool(&pool), mValue(&value) {}
+
+            SoundBufferSharedRef use() noexcept { return SoundBufferSharedRef(*mPool, *mValue); }
+
+            Sound_Buffer* get() const noexcept { return mValue; }
+
+            Sound_Buffer* operator->() const noexcept { return mValue; }
+
+        private:
+            SoundBufferPool* mPool = nullptr;
+            Sound_Buffer* mValue = nullptr;
+    };
```

Comment:
What's the point of this class? `Sound_Buffer` objects should remain valid for the lifetime of the sound manager, and are automatically cleared with it. It's the internal handle, which stores the actual sample data for the output backend, that gets loaded and unloaded as needed.

Reply:
`Sound_Buffer` doesn't know what pool does it belong to. This type provides this information. Also it represents a state of `Sound_Buffer` when it's already allocated but sound is not playing yet. Between [this](https://github.com/OpenMW/openmw/blob/94b9d8f67af7a35a0a6813c85f4e11d0fcfe20ee/apps/openmw/mwsound/soundmanagerimp.cpp#L509) and [this](https://github.com/OpenMW/openmw/blob/94b9d8f67af7a35a0a6813c85f4e11d0fcfe20ee/apps/openmw/mwsound/soundmanagerimp.cpp#L529) line.

And I don't understand what do you suggest instead. I can move this into `Sound_Buffer` since it already knows about iterator from the pool.

> and are automatically cleared with it

Before this change it's handled by `SoundManager` with code like:
```c++
if (sfx->mUses-- == 1)
    mUnusedBuffers.push_front(sfx);
```
and
```c++
if (sfx->mUses++ == 0)
{
    SoundList::iterator iter = std::find(mUnusedBuffers.begin(), mUnusedBuffers.end(), sfx);
    if(iter != mUnusedBuffers.end())
        mUnusedBuffers.erase(iter);
}
```
that is copy-pasted over multiple places. I don't see why RAII can't be used for internal handles to avoid code duplication and make a usage of `Sound_Buffer` objects safer.

Reply:
> `Sound_Buffer` doesn't know what pool does it belong to.

It doesn't need to, since it exists for the life of its pool/allocating container. And when the pool is destroyed, nothing should have a `Sound_Buffer`.

> Before this change it's handled by SoundManager with code like:
> ...
> and
> ...
> that is copy-pasted over multiple places. I don't see why RAII can't be used for internal handles to avoid code duplication and make a usage of `Sound_Buffer` objects safer.

Private methods can be added to `SoundManagerImpl` to avoid the duplication. As it is, `mUnusedBuffers` only exists to expediate unloading of least-recently-used samples buffers. And managing which sample buffers to unload, when, and how, seems like the job of the sound manager to me, rather than the pool for `Sound_Buffer`s (which holds the sample buffers). `SoundBufferPool`, `SoundBufferRef`, and `SoundBufferSharedRef` seems way too over-designed to me and is putting management work where it shouldn't be.

Reply:
`SoundManagerImpl` is overwhelmed with logic from a selection of the next music track from a playlist to play to a reference counting. I don't know what are the benefits of doing this. Reference counting for a `Sound_Handle` and playing/stopping sounds are different responsibilities. Sound manager still manages when to unload the sound by `SoundBufferSharedRef ` destruction.

Reply:
> `SoundManagerImpl` is overwhelmed with logic from a selection of the next music track from a playlist to play to a reference counting.

The next music track/playlist handling can certainly go into a sub-manager of sorts. I believe the intention is to eventually allow for mod-defined playlists. The sound manager just needs to make sure the right playlist is being used, and ensure a track from that playlist plays. What's currently there is an attempt to randomize tracks while avoiding the same track playing twice, basically a fallback dynamic playlist creator.

The reference counting in `Sound_Buffer` is so the sound manager knows when a sound buffer is no longer used, and becomes the next candidate for unloading after any previously-made-unused sounds when the cache fills up. On its own, it doesn't do anything except to indicate 'this is now unused' or 'this is now used'. Since the sound manager already handles starting and stopping sounds, and handles the output backend, it's the most appropriate place to say when a sound is no longer in use or is now in use, and to load and unload sounds from the backend as needed.